### ISSUE TITLE
Allow parenthesized element expressions.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#780]
Currently list literal element expressions are either `if ...` or `for ...` or plain expressions. We do not allow parentheses around an `if ...` element, so the following:
```dart
var l = [ if (cond1) (if (cond2) 0) else if (cond3) 1 ];
```
is not allowed. That is inconvenient because we also do not have a "no element" element that is briefer than `...[]`, so the way to write the above logic today is:
```dart
var l = [ if (cond1) if (cond2) 0 else ...[] else if (cond3) 1 ];
```

We should consider allowing parentheses around all element expressions. The only place it actually matters is delimiting an `if` with no else, but that is an annoying surprise in the current grammar.

**Comment by munificent** [comment]
That's an interesting edge case I hadn't considered. I think we could probably allow this.

> so the way to write the above logic today is:
> 
> ```dart
> var l = [ if (cond1) if (cond2) 0 else ...[] else if (cond3) 1 ];
> ```

The other, simpler, option is:

```dart
var l = [ if (cond1) ...[if (cond2) 0] else if (cond3) 1 ];
```

Basically, `...[` and `]` *are* the grouping characters for elements.

**Comment by eernstg** [comment]
It should just work: the following rule can be used in Dart.g without breaking anything in language_2:

```antlr
element
    : expressionElement
    | mapEntry
    | spreadElement
    | ifElement
    | forElement
    | '(' element ')'
    ;
```

**Comment by munificent** [comment]
> It should just work: the following rule can be used in Dart.g without breaking anything in `language_2`

Good to know.

I'm also slightly worried about it preventing other future syntax changes. (For example, you could imagine wanting to support `(for (var x = 0; x < 100; x++) x)` expression syntax as a way to create a lazy Iterable, similar to generator expressions in Python.

**Comment by tatumizer** [comment]
I wonder why not take a more obvious (and more consistent with the rest of the language) route with `{...}`:
```dart
// instead of this:
var l = [ if (cond1) if (cond2) 0 else ...[] else if (cond3) 1 ];
//write this:
var l = [ if (cond1) { if (cond2) 0 } else if (cond3) 1 ];
// instead of this
var m = [ for (int i=0; i<10; i++) ...[i, i*3] ];
// write this:
var m = [ for (int i=0; i<10; i++) {i, i*3} ];
```
That is, inside `{...}`, every emitted value gets inserted into the list. We can even express empty blocks if we feel so inclined:
```dart
var m = [ if (c) if (c1) 15 else {} else 10 ];
```
The only objection was that `{...}` collides with the sets literal and is therefore ambiguous. But set literals are rare, and in the rare cases they really occur in the code, they can be easily disambiguated:
```dart
var m = [ for (int i=0; i<10; i++) <int>{i, i*3} ];
```
(I believe set literals should have been designed with a required type parameter from the start, otherwise, they look weird).

However, the solution proposed in this ER is not perfect either.
```dart
var l = [ if (cond1) (if (cond2) 0) else if (cond3) 1 ];
// but I need 2 values for cond1 (conditional zero and unconditional 2), can I write this?
var l = [ if (cond1) (if (cond2) 0, 2) else if (cond3) 1 ];
```
Apparently, two values are not allowed. But if they ARE allowed, then why not expand this syntax so that the following is valid, too:
```dart
var l = [ if (cond1) (if (cond2) 0, 1) else if (cond3) 1 ];
var m = [ for (int i=0; i<10; i++) (i, i*3) ];
```
That is, we get exactly the same thing as I proposed above, just instead of `{...}` we use `(...)`. Which of these 2 variants is better, I don't know, but each of them provides a more general solution to the problem.

**Comment by lrhn** [comment]
Why not:
```dart
var l = [ if (cond1) { if (cond2) 0 } else if (cond3) 1 ];
```
Because `{ if (cond2) 0 }` is a set literal. Using braces for group as well as for creating set literals gives us an ambiguous grammer.

Parentheses are usually used for grouping expressions and braces for statements. "Collection elements" feels closer to expressions than statements - they introduce something into the place where they occur.

**Comment by tatumizer** [comment]
> Parentheses are usually used for grouping expressions and braces for statements. "Collection elements" feel closer to expressions than statements - they introduce something into the place where they occur.

My takeaway from this: 1) we don't like Braces for grouping 2) we like Parentheses for grouping
So, this is an example of the expression we like: `[ for (int i=0; i<10; i++) (i, i*3) ];` Right? Wrong?

**Comment by lrhn** [comment]
Wrong?

We like braces for grouping statements. We like parentheses for grouping *sub*-expressions. We don't generally group multiple expressions, those mainly exist lists (parameter lists, collection literals), and they are comma separated instead and the braces are delimiters, not grouping/precedence-operators.

The problem we're trying to solve here is precedence among collection elements.
Collection elements are ... something in-between statements and expressions. I think of them as closer to expressions because they have a value. Some have *more* than one value, but that's just the a list-monadic expression. Statements, on the other hand, do not. Even if I didn't read it as a set literal, I wouldn't think `[for (var x in xs) { x }]` as evaluating `{x}` for its value, that's not what braces read like to me. So, I'd prefer `[for (var x in xs) (x)]` (which already works as written, it just doesn't allow `x` to be replaced by a `for` or `if` element).

If the goal was to allow *multiple* elements inside an `if` or `for`, something different, I *might* actually prefer braces.
So, `[ for (int i=0; i<10; i++) {i, i*3} ]` could look better to me for *that* goal - not sure it does, neither that or using `(i, i*3)` are perfect, but both are reasonable choices. Using braces still has the set-literal issue, though.
For now I'll just use `...[i, i*3]` instead, and hope the compiler will be smart enough to recognize that.

I know it's all "look and feel" arguments, so clearly subjective. 😞 

(And, whoops, interpolations use `${...}` as an expression delimiter. Probably just because they stand out more, which is *also* important. Something like `"foo bar $(baz) qux (or fifi)"` is too easy to read as the parentheses being part of the text.)

**Comment by tatumizer** [comment]
(After a bit of research, came up with a plausible conjecture: string interpolation syntax is inherited from ksh. Ksh has both `${x}` and `$(x)`, but they mean different things - the former is "our" interpolation, but the latter means "execute command x, capture its output and insert it here". Everybody else just copied `${...}` from there).

**Comment by lrhn** [comment]
I agree that some Unix shell (or Perl, I think awk used `$(...)`) was likely the origin of `$`-interpolation. It just "looked right" to the original designers (and I don't disagree, I think it reads better than `$(foo)` would, because braces don't usually occur in text. Braces look awful in JSON literals, though, so maybe we should have allowed both). 

C# also uses braces, just without the `$` (or rather, with the `$` outside the string: `$"ab{foo}cd"`).
